proxy/envoy/discovery.go

```diff
@@ -15,24 +15,42 @@
 package envoy
 
 import (
+	"encoding/json"
 	"fmt"
 	"net/http"
 	"net/http/pprof"
 	"strconv"
+	"sync"
 
 	restful "github.com/emicklei/go-restful"
 	"github.com/golang/glog"
+	"github.com/golang/protobuf/proto"
 
 	proxyconfig "istio.io/api/proxy/v1/config"
 	"istio.io/manager/model"
 )
 
 // DiscoveryService publishes services, clusters, and routes for all proxies
 type DiscoveryService struct {
-	services model.ServiceDiscovery
-	config   *model.IstioRegistry
-	mesh     *proxyconfig.ProxyMeshConfig
-	server   *http.Server
+	services   model.ServiceDiscovery
+	controller model.Controller
+	config     *model.IstioRegistry
+	mesh       *proxyconfig.ProxyMeshConfig
+	server     *http.Server
+
+	disableCache bool
+	cacheMu      sync.Mutex
```

Comment:
s/cacheMu/cacheRWLock/ ?
Also, I suggest using sync.RWMutex instead of plain mutex.